File: projects/poker/go/poker.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"strings"
)

func main() {
	fmt.Println("\n*** P O K E R  H A N D  A N A L Y Z E R ***\n\n")

	var hand []*Card

	args := os.Args[1:]

	if len(args) == 0 {
		fmt.Println("*** USING RANDOMIZED DECK OF CARDS ***")
		hand = generateDeck()
	}

	if len(args) == 1 {
		filename := args[0]
		fmt.Println("*** USING TEST DECK ***\n")
		fmt.Println("*** File:", filename)
		hand = processFile(filename)
	}

	if len(args) == 0 {
		fmt.Println("*** Here is the hand")
		printHand(hand)
	}

	hands := generateCombinations(hand)
	printWin(hands)
}

func generateDeck() []*Card {
	suits := [4]string{"D", "C", "H", "S"}
	ranks := [13]string{"A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"}
	allCards := []*Card{}

	for _, s := range suits {
		for _, r := range ranks {
			allCards = append(allCards, NewCard(r, s))
		}
	}

	var deck [52]*Card

	for _, c := range allCards {
		pos := rand.Intn(52)
		for deck[pos] != nil {
			pos = (pos + 1) % 52
		}
		deck[pos] = c
	}

	fmt.Println("\n*** Shuffled 52 card deck:")
	for i, card := range deck {
		fmt.Print(card.toString(), " ")
		if (i+1)%13 == 0 {
			fmt.Println()
		}
	}
	fmt.Println()

	sixCardHand := []*Card{}
	for i := 0; i < 6; i++ {
		sixCardHand = append(sixCardHand, deck[i])
	}
	return sixCardHand
}

func printHand(hand []*Card) {
	fmt.Print(" ")
	for _, c := range hand {
		fmt.Print(c.toString(), " ")
	}
	fmt.Println("\n")
}

func generateCombinations(hand []*Card) []*Hand {
	hands := []*Hand{}
	cardIndexTracker := 0
	excludedIndex := 0

	for i := 0; i < len(hand); i++ {
		newHand := []*Card{}

		for j := 0; j < len(hand)-1; j++ {
			index := cardIndexTracker % len(hand)
			cardIndexTracker++
			newHand = append(newHand, hand[index])
			excludedIndex = (index + 1) % len(hand)
		}

		hands = append(hands, NewHand(newHand, hand[excludedIndex]))
	}

	fmt.Println("*** Hand Combinations...")
	for _, h := range hands {
		h.printHand()
	}
	fmt.Println()

	return hands
}

func printWin(hands []*Hand) {
	sortHands(hands)
	fmt.Println(" --- HIGH HAND ORDER ---")
	for _, h := range hands {
		h.printWithLabel()
	}
	fmt.Println()
}

func processFile(filename string) []*Card {
	content, _ := os.ReadFile(filename)

	line := strings.TrimSpace(string(content))
	tokens := strings.Split(line, ",")
	hand := []*Card{}
	duplicate := false
	var duplicateCard *Card

	for _, token := range tokens {
		cardStr := strings.TrimSpace(token)
		var rank string
		var suit string

		if len(cardStr) == 3 {
			rank = cardStr[:2]
			suit = cardStr[2:]
		} else {
			rank = cardStr[:1]
			suit = cardStr[1:]
		}

		newCard := NewCard(rank, suit)

		for _, c := range hand {
			if newCard.equals(c) {
				duplicate = true
				duplicateCard = c
			}
		}
		hand = append(hand, newCard)
	}

	if strings.HasPrefix(line, "10") {
		fmt.Println(line + "\n")
	} else {
		fmt.Println(" " + line + "\n")
	}

	if duplicate {
		fmt.Println("*** ERROR - DUPLICATED CARD FOUND IN DECK ***\n")
		fmt.Println("*** DUPLICATE:", duplicateCard.toString(), "***")
		os.Exit(1)
	}

	return hand
}

func highCardTieBreaker(h1, h2 *Hand) bool {
	for i := len(h1.cards) - 1; i >= 0; i-- {
		if h1.cards[i].rankValue() > h2.cards[i].rankValue() {
			return false
		} else if h1.cards[i].rankValue() < h2.cards[i].rankValue() {
			return true
		} else {
			if h1.cards[i].suitValue() > h2.cards[i].suitValue() {
				return false
			} else if h1.cards[i].suitValue() < h2.cards[i].suitValue() {
				return true
			}
		}
	}
	return false
}

func pairTieBreaker(h1, h2 *Hand) bool {
	var pair1 []*Card
	var pair2 []*Card
	var other1 []*Card
	var other2 []*Card

	for _, c := range h1.cards {
		if c.rankValue() == h1.highestCard.rankValue() {
			pair1 = append(pair1, c)
		} else {
			other1 = append(other1, c)
		}
	}
	for _, c := range h2.cards {
		if c.rankValue() == h2.highestCard.rankValue() {
			pair2 = append(pair2, c)
		} else {
			other2 = append(other2, c)
		}
	}

	for i := len(pair1) - 1; i >= 0; i-- {
		if pair1[i].rankValue() > pair2[i].rankValue() {
			return false
		} else if pair1[i].rankValue() < pair2[i].rankValue() {
			return true
		} else {
			if pair1[i].suitValue() > pair2[i].suitValue() {
				return false
			} else if pair1[i].suitValue() < pair2[i].suitValue() {
				return true
			}
		}
	}

	for i := len(other1) - 1; i >= 0; i-- {
		if other1[i].rankValue() > other2[i].rankValue() {
			return false
		} else if other1[i].rankValue() < other2[i].rankValue() {
			return true
		} else {
			if other1[i].suitValue() > other2[i].suitValue() {
				return false
			} else if other1[i].suitValue() < other2[i].suitValue() {
				return true
			}
		}
	}
	return false
}

func threeTieBreaker(h1, h2 *Hand) bool {
	var trip1 []*Card
	var trip2 []*Card
	var other1 []*Card
	var other2 []*Card

	for _, c := range h1.cards {
		if c.rankValue() == h1.highestCard.rankValue() {
			trip1 = append(trip1, c)
		} else {
			other1 = append(other1, c)
		}
	}
	for _, c := range h2.cards {
		if c.rankValue() == h2.highestCard.rankValue() {
			trip2 = append(trip2, c)
		} else {
			other2 = append(other2, c)
		}
	}

	for i := len(trip1) - 1; i >= 0; i-- {
		if trip1[i].rankValue() > trip2[i].rankValue() {
			return false
		} else if trip1[i].rankValue() < trip2[i].rankValue() {
			return true
		} else {
			if trip1[i].suitValue() > trip2[i].suitValue() {
				return false
			} else if trip1[i].suitValue() < trip2[i].suitValue() {
				return true
			}
		}
	}

	for i := len(other1) - 1; i >= 0; i-- {
		if other1[i].rankValue() > other2[i].rankValue() {
			return false
		} else if other1[i].rankValue() < other2[i].rankValue() {
			return true
		} else {
			if other1[i].suitValue() > other2[i].suitValue() {
				return false
			} else if other1[i].suitValue() < other2[i].suitValue() {
				return true
			}
		}
	}
	return false
}

func fourTieBreaker(h1, h2 *Hand) bool {
	if h1.highestCard.rankValue() > h2.highestCard.rankValue() {
		return false
	} else if h1.highestCard.rankValue() < h2.highestCard.rankValue() {
		return true
	} else {
		return h1.highestCard.suitValue() < h2.highestCard.suitValue()
	}
}

func twoPairTieBreaker(h1, h2 *Hand) bool {
	var pairs1 []*Card
	var pairs2 []*Card

	pairs1 = append(pairs1, h1.highestCard)
	pairs2 = append(pairs2, h2.highestCard)

	for i := 0; i < len(h1.cards); i++ {
		if !h1.cards[i].equals(h1.highestCard) {
			pairs1 = append(pairs1, h1.cards[i])
		}
		if !h2.cards[i].equals(h2.highestCard) {
			pairs2 = append(pairs2, h2.cards[i])
		}
	}

	for i := len(pairs1) - 1; i >= 0; i-- {
		if pairs1[i].rankValue() > pairs2[i].rankValue() {
			return false
		} else if pairs1[i].rankValue() < pairs2[i].rankValue() {
			return true
		} else if pairs1[i].suitValue() > pairs2[i].suitValue() {
			return false
		} else if pairs1[i].suitValue() < pairs2[i].suitValue() {
			return true
		}
	}
	return false
}

func straightTieBreaker(h1, h2 *Hand) bool {
	return highCardTieBreaker(h1, h2)
}
func straightFlushTieBreaker(h1, h2 *Hand) bool{
	return highCardTieBreaker(h1, h2)
}
func flushTieBreaker(h1, h2 *Hand) bool {
	return highCardTieBreaker(h1, h2)
}
func fullHouseTieBreaker(h1, h2 *Hand) bool {
	return threeTieBreaker(h1, h2)
}

func sortHands(hands []*Hand) {
	swapped := true
	for swapped {
		swapped = false
		for i := 0; i < len(hands)-1; i++ {
			h1 := hands[i]
			h2 := hands[i+1]

			if h1.handValue < h2.handValue {
				hands[i], hands[i+1] = hands[i+1], hands[i]
				swapped = true
			} else if h1.handValue == h2.handValue {
				needSwap := false

				if h1.label == "High Card" && h2.label == "High Card" {
					needSwap = highCardTieBreaker(h1, h2)
				} else if h1.label == "Pair" && h2.label == "Pair" {
					needSwap = pairTieBreaker(h1, h2)
				} else if h1.label == "Four of a Kind" && h2.label == "Four of a Kind" {
					needSwap = fourTieBreaker(h1, h2)
				} else if h1.label == "Three of a Kind" && h2.label == "Three of a Kind" {
					needSwap = threeTieBreaker(h1, h2)
				} else if h1.label == "Full House" && h2.label == "Full House" {
					needSwap = fullHouseTieBreaker(h1, h2)
				} else if h1.label == "Straight Flush" && h2.label == "Straight Flush" {
					needSwap = straightFlushTieBreaker(h1, h2)
				} else if h1.label == "Flush" && h2.label == "Flush" {
					needSwap = flushTieBreaker(h1, h2)
				} else if h1.label == "Straight" && h2.label == "Straight" {
					needSwap = straightTieBreaker(h1, h2)
				} else if h1.label == "Two Pair" && h2.label == "Two Pair" {
					needSwap = twoPairTieBreaker(h1, h2)
				}

				if needSwap {
					hands[i], hands[i+1] = hands[i+1], hands[i]
					swapped = true
				}
			}
		}
	}
}

File: projects/poker/c++/poker.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <random>
#include <fstream>
#include <sstream>
#include <cstdlib>
#include "card.h"
#include "hand.h"

using namespace std;

vector<Card> generateDeck();
void printHand(vector<Card> hand);
vector<Hand> generateCombinations(vector<Card> hand);
void printWin(vector<Hand> hands);
vector<Card> processFile(const string& filename);
bool highCardTieBreaker(Hand& h1, Hand& h2);
void sortHands(vector<Hand>& hands);
bool pairTieBreaker(Hand& h1, Hand& h2);
bool fourTieBreaker(Hand& h1, Hand& h2);
bool threeTieBreaker(Hand &h1, Hand& h2);
bool straightFlushTieBreaker(Hand& h1, Hand& h2);
bool straightTieBreaker(Hand& h1, Hand& h2);
bool fullHouseTieBreaker(Hand& h1, Hand& h2);
bool flushTieBreaker(Hand& h1, Hand& h2);

int main(int argc, char* argv[])
{
	cout << "\n*** P O K E R  H A N D  A N A L Y Z E R ***\n\n\n";

	vector<Card> hand;	// vector to hold the 6 card hand

	if (argc == 1)
	{
		cout << "*** USING RANDOMIZED DECK OF CARDS ***\n";
		hand = generateDeck();	// generates the decks and returns the 6 card hand
	}
	if (argc == 2)
	{
		string filename = argv[1];	
		cout << "*** USING TEST DECK ***\n\n";
		cout << "*** File: " << filename << endl;
		hand = processFile(filename);	
	}

	if (argc == 1)
	{
		cout << "*** Here is the hand\n";
		printHand(hand);	// print the 6 card hand
	}

	vector<Hand> hands = generateCombinations(hand);

	printWin(hands);

	return 0;
}

vector<Card> generateDeck()
{
	
	vector<string> suits = {"D","C","H","S"};
	vector<string> ranks = {"A","2","3","4","5","6","7","8","9","10","J","Q","K"};
	vector<Card> allCards;

	// generate all 52 cards
	for (string suit : suits)
	{
		for (string rank : ranks)
		{
			allCards.push_back(Card(rank, suit));
		}
	}

	vector<Card> deck(52); // vector to hold 52 cards randomized

	srand(time(0)); // seed the random number generator at current time

	// assign each card to a place in the deck
	for (Card c : allCards)	
	{
		int pos = rand() % 52;	// get the position it should be in

		while (!deck[pos].rank.empty())	// fill the deck vector based on empty positions
		{
			pos = (pos + 1) % 52;
		}
		deck[pos] = c;	
	}

	// print the shuffled 52 card deck
	cout << "\n*** Shuffled 52 card deck:\n";
	for (int i = 0; i < deck.size(); i++)
	{
		cout << deck[i].toString() << " ";
		if ((i+1) % 13 == 0)
			cout << "\n";
	}
	cout << endl;

	// get the first six cards in the shuffled deck
	vector<Card> hand;
	for (int i = 0; i < 6; i++)
		hand.push_back(deck[i]);

	return hand;
}

void printHand(vector<Card> hand)
{
	// prints each card in the six card hand
	cout << " ";
	for (Card card : hand)
		cout << card.toString() << " ";
	cout << "\n\n";
}

vector<Hand> generateCombinations(vector<Card> hand)
{
	vector<Hand> hands;

	int cardIndexTracker = 0;
	int excludedIndex = 0;

	// make the combinations by going through all the cards in order
	// make the excluded card the next one in the order
	// keep track of the cardIndex, so the next card will be the first in the next hand
	for (int i = 0; i < hand.size(); i++)
	{
		vector<Card> newHand;
		for (int j = 0; j < hand.size() - 1; j++)
		{
			int index = cardIndexTracker % hand.size();	// make sure the index does not go over 5
			cardIndexTracker++;		
			newHand.push_back(hand[index]);	
			excludedIndex = (index + 1) % hand.size(); //excluded card will be the next card after the loop ends
		}
		
		hands.push_back(Hand(newHand, hand[excludedIndex]));
		
	}

	cout << "*** Hand Combinations...\n";
	for (Hand currHand : hands)
	{
		currHand.toString();
	}
	cout << endl;

	return hands;
}

void printWin(vector<Hand> hands)
{
	sortHands(hands);
	cout << " --- HIGH HAND ORDER ---\n";
	for (Hand hand : hands)
		hand.printWithLabel();
	cout << endl;
}

vector<Card> processFile(const string& filename)
{
	ifstream file(filename);	
	vector<Card> hand;
	string line;
	bool duplicate = false;
	Card duplicateCard;

	if (getline(file, line)) 
	{
		stringstream ss(line);	
		string cardStr;

		while (getline(ss, cardStr, ',')) // split on comma
		{
			// trim leading/trailing spaces
			cardStr.erase(0, cardStr.find_first_not_of(' '));
			cardStr.erase(cardStr.find_last_not_of(' ') + 1);

			string rank, suit;
			if (cardStr.size() == 3)	// case for "10" 
			{
				rank = cardStr.substr(0, 2);
				suit = cardStr.substr(2, 1);
			}
			else if (cardStr.size() == 2) 
			{
				rank = cardStr.substr(0, 1);
				suit = cardStr.substr(1, 1);
			}
			Card newCard = Card(rank,suit);
			for (Card c : hand)
			{
				if (newCard == c)
				{
					duplicate = true;
					duplicateCard = c;		
				}
			}
			hand.push_back(newCard);
		}
	}
	
	if (line.substr(0, 2) == "10")
		cout << line << endl << endl;
	else
		cout << " " << line << endl << endl;
	
	if (duplicate)
	{
		cout << "*** ERROR - DUPLICATED CARD FOUND IN DECK ***\n\n";
		cout << "*** DUPLICATE: " + duplicateCard.toString() + " ***\n";
		exit(1);
	}
	return hand;
}


bool twoPairTieBreaker(Hand& h1, Hand& h2)
{
	vector<Card> pairs1;
	vector<Card> pairs2;
	
	// highest card is the kicker card
	// push it to the front to compare last
	pairs1.push_back(h1.highestCard); 
	pairs2.push_back(h2.highestCard);

	// push the pairs into vector
	for (int i = 0; i < h1.cards.size(); i++)
	{
		if (!(h1.cards[i] == h1.highestCard))
			pairs1.push_back(h1.cards[i]);

		if (!(h2.cards[i] == h2.highestCard))
			pairs2.push_back(h2.cards[i]);
	}

	// compare ranks then suits 
	for (int i = pairs1.size() - 1; i >= 0; i--)
	{
		if (pairs1[i].rankValue() > pairs2[i].rankValue())
			return false;
		if (pairs1[i].rankValue() < pairs2[i].rankValue())
		{
			swap(h1,h2);
			return true;
		}
		if (pairs1[i].suitValue() > pairs2[i].suitValue())
			return false;
		if (pairs1[i].suitValue() < pairs2[i].suitValue())
		{
			swap(h1, h2);
			return true;
		}
	}

	
	return false;

}

bool highCardTieBreaker(Hand& h1, Hand& h2)
{
	// compare the cards rank and suit from back to front
	for (int i = h1.cards.size() - 1; i >= 0; i--)
	{
		if (h1.cards[i].rankValue() > h2.cards[i].rankValue())
		{
			return false;
		}
		else if (h1.cards[i].rankValue() < h2.cards[i].rankValue())
		{
			swap(h1, h2);
			return true;
		}
		else
		{
			if (h1.cards[i].suitValue() > h2.cards[i].suitValue())
			{
				return false;
			}
			else if (h1.cards[i].suitValue() < h2.cards[i].suitValue())
			{
				swap(h1, h2);
				return true;
			}
		}
	}
	return false;
}

bool threeTieBreaker(Hand& h1, Hand& h2)
{
	vector<Card> trip1;
	vector<Card> trip2;
	vector<Card> other1;
	vector<Card> other2;

	// push the three of a kind into the trip vector
	for (int i = 0; i < h1.cards.size(); i++)
	{
		if (h1.cards[i].rankValue() == h1.highestCard.rankValue())
			trip1.push_back(h1.cards[i]);
		else
			other1.push_back(h1.cards[i]);

		if (h2.cards[i].rankValue() == h2.highestCard.rankValue())
                        trip2.push_back(h2.cards[i]);
                else
                        other2.push_back(h2.cards[i]);

	}

	// check the three of a kind rank and suit
	for (int i = trip1.size() - 1; i >= 0; i--)
	{
		if (trip1[i].rankValue() > trip2[i].rankValue())
			return false;
		else if(trip1[i].rankValue() < trip2[i].rankValue())
		{
			swap(h1,h2);
			return true;
		}
		else
		{	
			if (trip1[i].suitValue() > trip2[i].suitValue())
				return false;
			if(trip1[i].suitValue() < trip2[i].suitValue())
			{
				swap(h1, h2);
				return true;
			}
		}
	}

	// check the other two cards rank and suit
	for (int i = other1.size() - 1; i >= 0; i--)
	{
		if (other1[i].rankValue() > other2[i].rankValue())
                        return false;
                else if(other1[i].rankValue() < other2[i].rankValue())
                {
                        swap(h1,h2);
                        return true;
                }
                else
                {
                        if (other1[i].suitValue() > other2[i].suitValue())
                                return false;
                        if(other1[i].suitValue() < other2[i].suitValue())
                        {
                                swap(h1, h2);
                                return true;
                        }
                }
	}
	return false;
}

bool fourTieBreaker(Hand& h1, Hand& h2)
{
	// highest card for four of a kind contains the kicker card, 
	// only have to evaluate the rank and suit of kicker card because four kind will always be same
	
	if (h1.highestCard.rankValue() > h2.highestCard.rankValue())
		return false;
	else if (h1.highestCard.rankValue() < h2.highestCard.rankValue())
	{
		swap(h1,h2);
		return true;
	}
	else
	{
		if (h1.highestCard.suitValue() < h2.highestCard.suitValue())
		{
			swap(h1,h2);
			return true;
		}
	}
	return false;
}	

bool pairTieBreaker(Hand &h1, Hand& h2)
{
	vector<Card> pair1;
	vector<Card> pair2;
	vector<Card> other1;
	vector<Card> other2;
	
	// add the pair into the pair array and other cards into other array
	for (int i = 0; i < h1.cards.size(); i++)
	{
		if (h1.cards[i].rankValue() == h1.highestCard.rankValue())
			pair1.push_back(h1.cards[i]);
		else
			other1.push_back(h1.cards[i]);

		if (h2.cards[i].rankValue() == h2.highestCard.rankValue())
                        pair2.push_back(h2.cards[i]);
                else
                        other2.push_back(h2.cards[i]);

	}

	// compare suit and rank of pair first
	for (int i = pair1.size() - 1; i >= 0; i--)
	{
		if (pair1[i].rankValue() > pair2[i].rankValue())
			return false;
		else if(pair1[i].rankValue() < pair2[i].rankValue())
		{
			swap(h1,h2);
			return true;
		}
		else
		{	
			if (pair1[i].suitValue() > pair2[i].suitValue())
				return false;
			if(pair1[i].suitValue() < pair2[i].suitValue())
			{
				swap(h1, h2);
				return true;
			}
		}
	}

	//compare suit and rank of other three cards
	for (int i = other1.size() - 1; i >= 0; i--)
	{
		if (other1[i].rankValue() > other2[i].rankValue())
                        return false;
                else if(other1[i].rankValue() < other2[i].rankValue())
                {
                        swap(h1,h2);
                        return true;
                }
                else
                {
                        if (other1[i].suitValue() > other2[i].suitValue())
                                return false;
                        if(other1[i].suitValue() < other2[i].suitValue())
                        {
                                swap(h1, h2);
                                return true;
                        }
                }
	}
	return false;
}

bool straightTieBreaker(Hand& h1, Hand& h2)
{
	// only one card will be different in straight tie breaker, so we can just do highCardTieBreaker
	return highCardTieBreaker(h1, h2);
}

bool straightFlushTieBreaker(Hand& h1, Hand& h2)
{
	// only one card will be different in straight flush tie breaker, so we can just do highCardTieBreaker
	return highCardTieBreaker(h1, h2);
}

bool flushTieBreaker(Hand& h1, Hand& h2)
{
	// only one card will be different in flush tie breaker, so we can just do highCardTieBreaker
	return highCardTieBreaker(h1, h2);
}

bool fullHouseTieBreaker(Hand& h1, Hand& h2)
{
	// same logic as three of a kind, with others1 and others2 being the pairs
	return threeTieBreaker(h1, h2);
}

void sortHands(vector<Hand>& hands) // bubble sort algorithm
{
	bool swapped;
	bool didSwap;

	do
	{
		swapped = false;

		for (int i = 0; i < hands.size() - 1; i++)
		{
			if (hands[i].handValue < hands[i+1].handValue)
			{
				swap(hands[i], hands[i+1]);
				swapped = true;
			}
			else if (hands[i].handValue == hands[i+1].handValue)
			{
				if (hands[i].label == "High Card" && hands[i+1].label == "High Card")
					didSwap = highCardTieBreaker(hands[i], hands[i+1]);

				else if (hands[i].label == "Pair" && hands[i+1].label == "Pair")
					didSwap = pairTieBreaker(hands[i], hands[i+1]);

				else if (hands[i].label == "Four of a Kind" && hands[i+1].label == "Four of a Kind")
					didSwap = fourTieBreaker(hands[i], hands[i+1]);

				else if (hands[i].label == "Three of a Kind" && hands[i+1].label == "Three of a Kind")
					didSwap = threeTieBreaker(hands[i], hands[i+1]);

				else if (hands[i].label == "Full House" && hands[i+1].label == "Full House")
					didSwap = fullHouseTieBreaker(hands[i], hands[i+1]);

				else if (hands[i].label == "Straight Flush" && hands[i+1].label == "Straight Flush")
					didSwap = straightFlushTieBreaker(hands[i], hands[i+1]);

				else if (hands[i].label == "Flush" && hands[i+1].label == "Flush")
					didSwap = flushTieBreaker(hands[i], hands[i+1]);
	
				else if (hands[i].label == "Straight" && hands[i+1].label == "Straight")
					didSwap = straightTieBreaker(hands[i], hands[i+1]);
				
				else if (hands[i].label == "Two Pair" && hands[i+1].label == "Two Pair")
					didSwap = twoPairTieBreaker(hands[i], hands[i+1]);

				if (didSwap)
					swapped = true;
			}
		}
	} while (swapped);
}

